x/incentives/keeper/distribute.go

```diff
@@ -308,29 +321,105 @@ func (k Keeper) AllocateAcrossGauges(ctx sdk.Context) error {
+func (k Keeper) syncGroupGaugeWeights(ctx sdk.Context, groupGauge types.GroupGauge) error {
+	if groupGauge.SplittingPolicy == types.Volume {
+		err := k.syncVolumeSplitGauge(ctx, groupGauge)
+		if err != nil {
+			return err
+		}
+	} else {
+		return types.UnsupportedSplittingPolicyError{GroupGaugeId: groupGauge.GroupGaugeId, SplittingPolicy: groupGauge.SplittingPolicy}
+	}
+
+	return nil
+}
+
+// syncVolumeSplitGauge syncs a group gauge according to volume splitting policy.
+// It mutates the passed in object and sets the updated value in state.
+// If there is an error, the passed in object is not mutated.
+//
+// It returns an error if:
+// - the splitting policy is not supported
+// - the volume for any linked pool is zero or cannot be found
+// - the cumulative volume for any linked pool has decreased (should never happen)
+func (k Keeper) syncVolumeSplitGauge(ctx sdk.Context, groupGauge types.GroupGauge) error {
+	totalWeight := sdk.ZeroInt()
+
+	// We operate on a deep copy of the given group gauge because we expect to handle specific errors quietly
+	// and want to avoid the scenario where the original group gauge is partially mutated in such cases.
+	updatedGroupGauge := types.GroupGauge{
+		GroupGaugeId: groupGauge.GroupGaugeId,
+		InternalGaugeInfo: types.InternalGaugeInfo{
+			TotalWeight:  groupGauge.InternalGaugeInfo.TotalWeight,
+			GaugeRecords: make([]types.InternalGaugeRecord, len(groupGauge.InternalGaugeInfo.GaugeRecords)),
+		},
+		SplittingPolicy: groupGauge.SplittingPolicy,
+	}
+
+	// Loop through gauge records and update their state to reflect new pool volumes
+	for i, gaugeRecord := range groupGauge.InternalGaugeInfo.GaugeRecords {
+		gauge, err := k.GetGaugeByID(ctx, gaugeRecord.GaugeId)
+		if err != nil {
+			return err
+		}
+
+		gaugeType := gauge.DistributeTo.LockQueryType
+		gaugeDuration := time.Duration(0)
+
+		if gaugeType == lockuptypes.NoLock {
+			// If NoLock, it's a CL pool, so we set the "lockableDuration" to epoch duration
+			gaugeDuration = k.GetEpochInfo(ctx).Duration
+		} else {
+			// Otherwise, it's a balancer pool so we set it to longest lockable duration
+			// TODO: add support for CW pools once there's clarity around default gauge type.
+			// Tracked in issue https://github.com/osmosis-labs/osmosis/issues/6403
+			gaugeDuration, err = k.pik.GetLongestLockableDuration(ctx)
+			if err != nil {
+				return err
+			}
+		}
+
+		// Retrieve pool ID using GetPoolIdFromGaugeId(gaugeId, lockableDuration)
+		poolId, err := k.pik.GetPoolIdFromGaugeId(ctx, gaugeRecord.GaugeId, gaugeDuration)
+		if err != nil {
+			return err
+		}
 
-		var coinsDistPerInternalGauge, coinsDistThisEpoch sdk.Coins
-		for _, coin := range remainCoins {
-			epochDiff := groupGauge.NumEpochsPaidOver - groupGauge.FilledEpochs
-			internalGaugeLen := len(groupGaugeObj.InternalIds)
+		// Get new volume for pool. Assert GTE gauge's weight
+		cumulativePoolVolume := k.pmk.GetOsmoVolumeForPool(ctx, poolId)
```

Comment:
Note to reviewer: this section would probably be best to review in split view. The entire function was re-implemented, so the line-by-line changes are not relevant